tensorflow/java/src/main/java/org/tensorflow/op/core/Constant.java

```diff
@@ -64,10 +48,24 @@
    * @param scope is a scope used to add the underlying operation.
    * @param shape the tensor shape.
    * @param data a buffer containing the tensor data.
+   * @return an integer constant
    * @throws IllegalArgumentException If the tensor shape is not compatible with the buffer
    */
-  public static Constant<Integer> create(Scope scope, long[] shape, IntBuffer data) {
-    try (Tensor<Integer> value = Tensor.create(shape, data)) {
+  public static Constant<Integer> create(Scope scope, Shape shape, IntBuffer data) {
```

Comment:
I'm sure I'm missing something, but why is the change from `long[]` to `Shape` needed?
`Shape` can represent partially known shapes (unknown rank, or some dimensions unknown). In the API so far we've been using `Shape` only in places where it is okay to have a partially known shape, while `long[]` where the shape is fully known (the assumption being that all elements of the array are >= 0).

In these `create()` functions, we want fully known shapes, right?

Reply:
Didn't quite follow the comment about `Shape.byteSize()`, but I'm guessing you mean utility methods like `numElements()`?

Auto generated operators accept `Shape` as arguments because the operations work with partially known shapes. `Constant` objects require a fully known shape, no?

I appreciate the interest in standardizing, but I'm not sold on it here just yet. Perhaps some sample code that we'd expect users to write where providing a `Shape` object would be better would make this clearer? Naively, I'm thinking that the type mismatch will make the developer think about the difference, vs. providing a `Shape` object and then observing a runtime error if `Shape` isn't fully specified.

Reply:
Yes, I meant `numElements()` here, sorry for the confusion. 

What I tried to explain is that manipulating primitive arrays as we do here can be restrictive when we want to extend the functionalities of the API. It could also lead to confusion that there is two way to represent the same thing and conversion from one to another can be required (I'll try to provide you an example of that).

Having a `Shape` object make sense to me as it is ubiquitous in TF domain. Just throwing ideas here : would it make sense to have a `FixedShape` class that extends from `Shape` to enforce that a shape must be known for some operations?

Reply:
To make sure I understand your proposal, you're suggesting that we use `Shape` instead of `long[]` everywhere (e.g., `Tensor.shape()` should return a `Shape`, and the factory functions should accept `Shape` instead of `long[]`), right?

You mention that the current separate "could lead to confusion since there are two ways to represent the same thing" - but, at least the way I was thinking about it, we represent them differently because they are *different* things :). The shape of a `Tensor` is concrete like the `.length`s of an array. While a `Shape` object is used to represent a pattern - and these patterns are consumed in specific contexts (like attributes of an operation). Hence, you can easily convert one way (`Shape.make()`), but not the other. 

Not a perfect analogy, but something along the lines of a `String` vs. `java.util.regex.Pattern`, or an `int` vs. a `BigInteger`. 

That said, perhaps I'm missing something. Some concrete examples of where this distinction is really hurtful will help. Or perhaps you can share some ideas on what kinds of flexibility are you thinking of for future usage?

Reply:
No, I think it's a good idea to keep enforcing only known shapes for some operations, such as `Tensor` or `Constant` creation. My last proposal was to replace the less-flexible primitive arrays by another concrete type (`FixedShape` or something similar) so we can easily retrieve properties of that shape.

For example, to retrieve the number of elements in a tensors, it is easier to do `tensor.shape().numElements()` than doing the computation manually from a `long[]` like the `Tensor` class is doing right now. Of course, we could have those utilities exposed as a static method of some final object, for instance `Shapes.numElements(tensor.shape())`, but I always had a personal preference for more object-oriented programming.

Just to give a quick example out-of-my-head-right-now (sorry if it's a bad one...) but let's say someone who would like to initialize his NxM tensor of floats to `1.0f` would benefit from this utility as well instead of iterating through shape dimensions and sizes himself.

Reply:
Close yes, I had in mind something like that, hiding completely the arrays:
```
public final class KnownShape implements Shape {
    public static KnownShape make(long...) { ... }
    ...
    public long numElements() { ... }
    ...
}

public final class Tensor {
    ...
    public KnownShape shape() { ... }
    ...
}

public final class Zeros<T> implements Op, Operand<T> {
   ...
   public static <T> Zeros<T> create(Scope scope, Class<T> type, KnownShape shape) { ... }
   ...
}
```
and so on. I just find it more clear that we are dealing with the "known" constraint of the shape and easier to provide utilities around shapes in general.

Not arguing here by the way, just find the discussion interesting :)

Reply:
The arrays (or at least the values of individual dimensions) will be useful since they have meaning (e..g, batch size, number of channels in an image, sequence length etc.). So, if we hide that, we'll want to add accessors for each dimension.

I guess there are some utility functions (like `numElements`, or perhaps `rank`, or `equals` overrides or something) that may be nice to have. However, I feel that at this stage having a class for this is overkill. Let's see some more usage / pain with `long[]` before we go that route

Reply:
>>> The arrays (or at least the values of individual dimensions) will be useful since they have meaning
Yes of course, `shape.size(int)` would still be present

>>> Let's see some more usage / pain with long[] before we go that route
Ok, I'll update the PR consequently, thanks

Reply:
To confirm, the plan is to revert the changes to `Constant.java` and `Shape.java`, right?

Reply:
The plan is to revert `Tensor` and to use shapes as `long[]` again in `Constant`. But I would keep the new scalar factories in `Constant` though (e.g. `ops.constant(1.0f)`)